arch/arm/core/aarch64/isr_wrapper.S

```diff
@@ -97,12 +97,34 @@ SECTION_FUNC(TEXT, _isr_wrapper)
 	sub	x2, x2, #1
 	str	x2, [x1, #_kernel_offset_to_nested]
 
-	/* Check if we need to context switch */
-	ldr	x2, [x1, #_kernel_offset_to_current]
-	ldr	x3, [x1, #_kernel_offset_to_ready_q_cache]
-	cmp	x2, x3
+	/*
+	 * Save _current in x17 because this can be changed by
+	 * z_get_next_switch_handle()
+	 */
+	ldr	x17, [x1, #_kernel_offset_to_current]
```

Comment:
This might be more complicated than necessary.  The intent of the "interrupted" argument to z_get_next_switch_handle() is that it's a switch handle constructed on interrupt **entry**[1] to contain the saved state of whatever got interrupted.   So once you have it (it doesn't have to be a thread pointer, but it works just fine if it is) you just leave it in a callee-save register across the ISR, then use it here.

Also a future note: in SMP contexts, there is no "_current" location in memory (because obviously every CPU has its own current thread).  The architecture is expected to store a current `struct _cpu` pointer (see `arch_curr_cpu()`) somewhere in the CPU outside the toolchain-visible register set (e.g. the FS segment register on x86, a MISC special register on xtensa, etc...) that it can be accessed easily, and this has a current field.

[1] Obviously there can be complexities introduced for optimization reasons.  For example you might save off only the caller-save registers on entry, and save the callee-save ones here only in the circumstance where the threads are different.  But conceptually, "interrupted" is something you should already have once the handler starts running.

Reply:
@andyross I'm reworking this PR but I still have one question.

In my current rework I'm using SP as switch handler as detailed in the `kernel_arch_interface.h` comment, and that's fine for the cooperative switching.

I'm a bit puzzled about the non-cooperative switching though. I'm using `z_get_next_switch_handle()` to get the handle of the next thread to be scheduled. Let's say that a new thread must be scheduled so I'm supposed to save the current context (basically my SP after pushing all the registers) into the `switch_handler` of the current task before this is switched out. How can I get access to the `switch_handler` without using the `_current` pointer? Should we go through `struct _cpu` as you pointed out in the previous message?